### PR TITLE
LLVM Flang 17

### PR DESCRIPTION
**Pull request by h-vetinari**
Carrying forward the work from #27 to LLVM 17. Expanding the [open points](https://github.com/conda-forge/flang-feedstock/pull/27#issuecomment-1613888196) from the last discussion into a TODO:

* [x] Fix windows default compilation (currently failing on missing symbol)
  * Looks like: `error LNK2019: unresolved external symbol __udivti3` due to a combination of missing compiler-rt support when using MSVC, and probably 128bit-support. 
  * Currently backporting the in-progress https://reviews.llvm.org/D154660, see also https://github.com/llvm/llvm-project/issues/63712
  * Seems to have worked in conjunction with the patch and https://github.com/conda-forge/flang-feedstock/pull/28/commits/a913048e452c9c7346382ab8858be05da8027ff1
* [x] Get rid of sysroot-hacks
  * Currently `{{ compiler('cxx') }}` in the test environment is necessary to properly find the sysroot on linux because of [this](https://github.com/conda-forge/clangdev-feedstock/issues/227); however the GCC-activation will set `CONDA_BUILD_SYSROOT`, which is then picked up through our [patched](https://github.com/conda-forge/clangdev-feedstock/blob/main/recipe/patches/0003-clang-add-conda-specific-env-var-CONDA_BUILD_SYSROOT.patch) clang.
  * Solved (on linux) by https://github.com/conda-forge/clangdev-feedstock/pull/231
* [x] `libflang_main` needs to depend on `libfortran_main_{{ cross_platform }}` (not the other way around)
  * ~we also need to put it on a path that's distinct per target triple to support e.g. having both x64 & aarch64 in the same environment (this will likely need patching of flang)~
  * Currently deprioritized, because the interaction with clang is [tricky](https://github.com/conda-forge/flang-feedstock/pull/28#issuecomment-1654882927), and we need cross-compilation less urgently than windows builds.
* [ ] windows activation
  * we'll probably need to do something like the clang-win activation [does](https://github.com/conda-forge/clang-win-activation-feedstock/blob/main/recipe/activate-clang_win-64.bat#L12) to set +/- the equivalent of `/MD`, which will also likely have to go into a separate feedstock
* [x] drop osx builds
  * we're can drop osx builds as soon as they become troublesome: one compiler per language & platform is enough (and in contrast to clang, which can target the gcc runtime, flang cannot target gfortran). We're still keeping linux because it's easier to debug for general problems, but will have to make it conflict with `libgfortran-ng`.

Potentially:
* [x] improve flang vs. compiler-rt setup (i.e. detection of the latter when using msvc; ideally avoid patching)
* [ ] decide whether to carry patch renaming `flang-new` to `flang`

Also worth noting: https://reviews.llvm.org/D154869 will reshuffle the runtime libs a lot, but this didn't land for LLVM 17.

### CODE REVIEW

**Comment by h-vetinari** [comment]
@isuruf, I'd like to drop `libfortran_main_{{ target }}` for now. The reason is that - due to the design of flang being based on the clang driver - it's being [picked up](https://github.com/llvm/llvm-project/blob/release/17.x/clang/lib/Driver/ToolChains/CommonArgs.cpp#L937-L948) in a path that belongs to clang, so developing this patch would be very painful as we'd have to build a full clang build each time.

It's also much less relevant based on the discussion that we don't need flang on linux/osx, so we don't have to worry about cross-compilation for aarch/ppc etc. The only point where I see this becoming relevant soon-ish is for win-arm64 support, but I'd like to cross that bridge when we get there, and not block flang on win-64 over that.

Finally, https://reviews.llvm.org/D154869 will overhaul a lot of the upstream integration of the runtime libs (including fortran_main), so we'd have to redo whatever patches we come up with anyway.

**Comment by h-vetinari** [comment]
```
LINK: command "%BUILD_PREFIX%\Library\bin\lld-link.exe /nologo @CMakeFiles\bbc.rsp /out:bin\bbc.exe /implib:lib\bbc.lib /pdb:bin\bbc.pdb /version:0.0 /machine:x64 -nostdlib -Xclang --dependent-lib=msvcrt -fuse-ld=lld -Wl,-defaultlib:%BUILD_PREFIX%/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib /STACK:10000000 /INCREMENTAL:NO /subsystem:console /MANIFEST /MANIFESTFILE:bin\bbc.exe.manifest" failed (exit code 1) with the following output:
lld-link: warning: ignoring unknown argument '-nostdlib'
lld-link: warning: ignoring unknown argument '-Xclang'
lld-link: warning: ignoring unknown argument '--dependent-lib=msvcrt'
lld-link: warning: ignoring unknown argument '-fuse-ld=lld'
lld-link: warning: ignoring unknown argument '-Wl,-defaultlib:%BUILD_PREFIX%/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib'
lld-link: error: undefined symbol: __udivti3
```

Obviously the symbol isn't found if the path to the library containing it is ignored 🤦

Not sure why this is happening, by now this is only using LLVM components.

**Comment by isuruf** [comment]
> Not sure why it's still necessary after your https://github.com/conda-forge/clangdev-feedstock/pull/231.

That's for linux, not windows.

> I'd like to rebase this to: https://github.com/conda-forge/flang-feedstock/commit/f685a8677b29332a77caa5dffae108ad4d8bd07a + https://github.com/conda-forge/flang-feedstock/commit/d7d1affb3cdc7564427425e9daec16d5af5bd253 + https://github.com/conda-forge/flang-feedstock/commit/09817516a108cbd3c614b7aa0cc6bbebc6a9f60a + a rerender (and get rid of all the remove-then-revert churn) - would you be alright with that?

Sure

**Comment by h-vetinari** [comment]
> > Not sure why it's still necessary after your [conda-forge/clangdev-feedstock#231](https://github.com/conda-forge/clangdev-feedstock/pull/231).
> 
> That's for linux, not windows.

You had added a test using `set "CONDA_BUILD_SYSROOT="  # [win]` in that PR, so I thought it also applies to windows...

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,24 +1,23 @@
...
 
 source:
-  url: https://github.com/xoviat/flang/archive/{{ flang_commit }}.tar.gz
-  sha256: {{ sha256 }}
+  url: https://github.com/llvm/llvm-project/releases/download/llvmorg-{{ version.replace(".rc", "-rc") }}/llvm-project-{{ version.replace(".rc", "rc") }}.src.tar.xz
+  sha256: 01b80625c131b8281dcf76a4e6395d10f45835c60e27357c92abf5a98c9a0337
+  patches:
+    # link compiler-rt libs (upstream's find_compiler_rt_library doesn't work correctly);
+    # probably related to https://github.com/llvm/llvm-project/issues/63286
+    - patches/0001-always-link-in-compiler-rt-builtins-on-windows-apple.patch
```
Comment (isuruf) [inline recipe/meta.yaml:13]: Is this still needed?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -68,28 +93,40 @@ outputs:
...
-        - llvm-openmp =={{ llvm_version }}
+        - sysroot_{{ target_platform }} ==2.17  # [linux]
+        - clangdev =={{ version }}
+        - llvm-openmp =={{ version }}
         - {{ pin_subpackage('libflang', exact=True) }}
+        - {{ pin_subpackage('libfortran-main', exact=True) }}
     test:
+      requires:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
```
Comment (isuruf) [inline recipe/meta.yaml:123]:
```suggestion
        - {{ compiler('c') }}       # [win]
        - {{ compiler('cxx') }}     # [win]
```

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml:13]: IIRC, that was necessary to build at all, not just to run. I can try commenting it out, but I'll do that after.

**Comment by h-vetinari** [comment]
Alright, the sysroot on linux now gets found without having `{{ compiler("cxx") }}` do its own activation. 🥳

**Comment by h-vetinari** [comment]
I've updated the todo list in the OP. The last big open question is how to do the windows compiler activation correctly. When we last spoke about this you mentioned we might need a separate feedstock. Is that necessary even if we're not doing the cross-compilation setup for now?

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml]: Well, perhaps that was only the case as of 16.0 (when I added that patch; I just rebased for 17), seems the build still passes without it 🥳

**Comment by h-vetinari** [comment]
@isuruf, what do you think we publish this to `llvm_rc`? That way it's hidden away from general use, but I could start testing it out in scipy, and we'll see which problems we'll run into anyway...?

**Comment by isuruf** [comment]
> @isuruf, what do you think we publish this to llvm_rc?

Sure.

>  Is that necessary even if we're not doing the cross-compilation setup for now?

Nope

**Comment by isuruf** [comment]
> decide whether to carry patch renaming flang-new to flang

Let's keep using `flang-new` or change to `llvm-flang` as Chris Lattner mentioned. I don't think it's good to name it flang just yet.

**Comment by h-vetinari** [comment]
OK, thanks!

Both those things (`llvm_rc` and `flang-new`) represent the state of this PR, so it sounds like we're good to merge?!

**Comment by h-vetinari** [comment]
Gentle ping @isuruf - OK for you if we push the button here?

**Comment by isuruf** [comment]
Yeah, sounds good.

**Pull request merged**